src/librustc/middle/ty.rs

```diff
@@ -1335,53 +1336,101 @@ pub enum BoundRegion {
+    /// definition and not a concrete use of it. To get the correct `TyEnum`
     /// from the tcx, use the `NodeId` from the `ast::Ty` and look it up in
-    /// the `ast_ty_to_ty_cache`. This is probably true for `ty_struct` as
+    /// the `ast_ty_to_ty_cache`. This is probably true for `TyStruct` as
     /// well.
-    ty_enum(DefId, &'tcx Substs<'tcx>),
-    ty_uniq(Ty<'tcx>),
-    ty_str,
-    ty_vec(Ty<'tcx>, Option<usize>), // Second field is length.
-    ty_ptr(mt<'tcx>),
-    ty_rptr(&'tcx Region, mt<'tcx>),
+    TyEnum(DefId, &'tcx Substs<'tcx>),
+
+    /// A structure type, defined with `struct`.
+    ///
+    /// See warning about substitutions for enumerated types.
+    TyStruct(DefId, &'tcx Substs<'tcx>),
+
+    /// `Box<T>`; this is nominally a struct in the documentation, but is
+    /// special-cased internally. For example, it is possible to implicitly
+    /// move the contents of a box out of that box, and methods of any type
+    /// can have type `Box<Self>`.
+    TyBox(Ty<'tcx>),
+
+    /// The pointee of a string slice. Written as `str`.
+    TyStr,
+
+    /// An array with the given length, or the pointee
+    /// of an array slice.  Written as `[T; n]`, or `[T]`.
+    /// FIXME: It probably makes sense to separate these.
+    TyArrayOrSlice(Ty<'tcx>, Option<usize>),
+
+    /// A raw pointer. Written as `*mut T` or `*const T`
+    TyRawPtr(mt<'tcx>),
+
+    /// A reference; a pointer with an associated lifetime. Written as
+    /// `&a mut T` or `&'a T`.
+    TyReference(&'tcx Region, mt<'tcx>),
+
+    /// If the def-id is Some(_), then this is the type of a specific
+    /// fn item. Otherwise, if None(_), it a fn pointer type.
+    ///
+    /// FIXME: Conflating function pointers and the type of a
+    /// function is probably a terrible idea; a function pointer is a
+    /// value with a specific type, but a function can be polymorphic
+    /// or dynamically dispatched.
+    TyBareFn(Option<DefId>, &'tcx BareFnTy<'tcx>),
+
+    /// A trait, defined with `trait`.
+    TyTrait(Box<TraitTy<'tcx>>),
 
-    // If the def-id is Some(_), then this is the type of a specific
-    // fn item. Otherwise, if None(_), it a fn pointer type.
-    ty_bare_fn(Option<DefId>, &'tcx BareFnTy<'tcx>),
+    /// The anonymous type of a closure. Used to represent the type of
+    /// `|a| a`.
+    TyClosure(DefId, &'tcx Substs<'tcx>),
 
-    ty_trait(Box<TyTrait<'tcx>>),
-    ty_struct(DefId, &'tcx Substs<'tcx>),
+    /// A tuple type.  For example, `(i32, bool)`.
+    TyTuple(Vec<Ty<'tcx>>),
 
-    ty_closure(DefId, &'tcx Substs<'tcx>),
+    /// The projection of an associated type.  For example,
+    /// `<T as Trait<..>>::N`.
+    TyProjection(ProjectionTy<'tcx>),
 
-    ty_tup(Vec<Ty<'tcx>>),
+    /// A type parameter; for example, `T` in `fn f<T>(x: T) {}
+    TyParam(ParamTy),
 
-    ty_projection(ProjectionTy<'tcx>),
-    ty_param(ParamTy), // type parameter
+    /// A placeholder type used during type-checking.
```

Comment:
I wouldn't call this a placeholder. We use "type variable" elsewhere in the compiler.